model/labels/labels.go

```diff
@@ -357,6 +357,31 @@ func (ls Labels) DropMetricName() Labels {
 	return ls
 }
 
+// DropMetricDeletedName returns Labels with "__deleted__name__" removed.
+func (ls Labels) DropMetricDeletedName() Labels {
+	for i, l := range ls {
+		if l.Name == DeletedMetricName {
+			if i == 0 { // Make common case fast with no allocations.
+				return ls[1:]
+			}
+			// Avoid modifying original Labels - use [:i:i] so that left slice would not
+			// have any spare capacity and append would have to allocate a new slice for the result.
+			return append(ls[:i:i], ls[i+1:]...)
+		}
+	}
+	return ls
+}
+
+// MarkMetricNameForDeletion returns Labels with "__name__" marked for deletion.
+func (ls Labels) MarkMetricNameForDeletion() Labels {
+	for i, l := range ls {
+		if l.Name == MetricName {
+			ls[i].Name = DeletedMetricName
```

Comment:
Assuming that `labels.Labels` is shared between samples, we have to consider it as immutable and cannot change the label value in place here.